binderhub/repoproviders.py

```diff
@@ -83,6 +91,141 @@ def get_build_slug(self):
         return '{user}-{repo}'.format(user='Rick', repo='Morty')
 
 
+class GitRepoProvider(RepoProvider):
+    """Bare bones git repo provider.
+
+    Users must provide a spec of the following form.
+
+    <url-escaped-namespace>/<resolved_ref>
+
+    eg:
+    https%3A%2F%2Fgithub.com%2Fjupyterhub%2Fzero-to-jupyterhub-k8s/f7f3ff6d1bf708bdc12e5f10e18b2a90a4795603
+
+    This provider is typically used if you are deploying binderhub yourself and you require access to repositories that
+    are not in one of the supported providers.
+    """
+
+    name = Unicode("Git")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        url, resolved_ref = self.spec.rsplit('/', 1)
+        self.repo = urllib.parse.unquote(url)
+        if not resolved_ref:
+            raise ValueError("`resolved_ref` must be specified as a query parameter for the basic git provider")
+        self.sha1_validate(resolved_ref)
+        self.resolved_ref = resolved_ref
+
+    @gen.coroutine
+    def get_resolved_ref(self):
+        return self.resolved_ref
+
+    def get_repo_url(self):
+        return self.repo
+
+    def get_build_slug(self):
+        parsed = urllib.parse.urlparse(self.repo)
+        parts = [parsed.netloc]
+        parts.extend(parsed.path.strip('/').split('/'))
```

Comment:
IMO, the right thing to do here is to simply return something like `url + hash`. The code in builder ensures that you are sized appropriately. Only counting netloc + path means that http / https will produce same image, which is not something we should do IMO.

Reply:
Does the builder also escape things that are problematic?

Reply:
@mariusvniekerk hah, there's a TODO to make it so: https://github.com/jupyterhub/binderhub/blob/master/binderhub/builder.py#L92 but right now it doesn't.

I think we should fix that TODO by calling into escapism.escape from there so we can solve it centrally in one place though rather than per provider.

Reply:
I made https://github.com/jupyterhub/binderhub/pull/298 for this!